clippy_lints/src/local_assigned_single_value.rs

```diff
@@ -0,0 +1,326 @@
+        let mut v = V {
+            body: mir,
+            cx,
+            map: mir.local_decls.iter().map(|_| LocalUsageValues::default()).collect(),
+        };
+        v.visit_body(mir);
+
+        for (local, i) in v.map.iter_enumerated() {
+            let LocalUsageValues {
+                usage,
+                mut_ref_acquired,
+            } = i;
+
+            if !mut_ref_acquired && eval_nested_locals_are_constant(&v.map, i)
+                && eval_local(&v.map, i)
+                && let Some(local_decl) = mir.local_decls.get(local)
+                    && let [dbg_info] = &*mir
+                        .var_debug_info
+                        .iter()
+                        .filter(|info| info.source_info.span == local_decl.source_info.span)
+                        .collect_vec()
+                // Don't handle function arguments.
+                && dbg_info.argument_index.is_none()
+                // Ignore anything from a procedural macro, or locals we cannot prove aren't
+                // temporaries
+                && let Some(snippet) = snippet_opt(cx, dbg_info.source_info.span)
+                && snippet.ends_with(dbg_info.name.as_str())
+            {
+                span_lint(
+                    cx,
+                    LOCAL_ASSIGNED_SINGLE_VALUE,
+                    usage.iter().map(|i| i.span).collect_vec(),
+                    "local only ever assigned single value",
+                );
+            }
+        }
+    }
+}
+
+type LocalUsageMap<'tcx> = IndexVec<mir::Local, LocalUsageValues<'tcx>>;
+
+/// Holds the data we have for the usage of a local.
+#[derive(Debug, Default)]
+struct LocalUsageValues<'tcx> {
+    /// Where and what this local is assigned.
+    usage: Vec<Spanned<LocalUsage<'tcx>>>,
+    /// Whether it's mutably borrowed, ever. We should not lint this.
+    mut_ref_acquired: bool,
+}
+
+#[derive(Debug)]
+enum LocalUsage<'tcx> {
+    /// A single `Scalar`, for stuff like `i32` or `bool`.
+    Scalar(Scalar),
+    /// Any number of `Scalar`s. This is used for handling arrays and tuples
+    Scalars(Vec<Scalar>),
+    /// When a `Local` depends on the value of another local
+    DependsOn(mir::Local),
+    /// When a `Local` depends on the value of another local by accessing any of its fields or
+    /// indexing
+    DependsOnWithProj(mir::Local, &'tcx PlaceElem<'tcx>),
+    /// Used when a local's assigned a value we cannot prove is constant.
+    NonConst,
+    /// This is always overwritten.
+    Pending,
+}
+
+struct V<'a, 'tcx> {
+    body: &'a mir::Body<'tcx>,
+    cx: &'a LateContext<'tcx>,
+    map: LocalUsageMap<'tcx>,
+}
+
+impl<'a, 'tcx> Visitor<'tcx> for V<'a, 'tcx> {
+    fn visit_assign(&mut self, place: &Place<'tcx>, rvalue: &Rvalue<'tcx>, loc: Location) {
+        let Self { body, cx, map } = self;
+        let Some(stmt) = body.stmt_at(loc).left() else {
+            return;
+        };
+
+        if stmt.source_info.span.from_expansion() {
```

Comment:
You can't just ignore assignments from macros. The most permissive thing you can do here is to block linting for any locals which are assigned in macros. Trying to lint macros will be pain if you want to avoid false positives.